Userland/Libraries/LibDisassembly/riscv64/IM.cpp

```diff
@@ -0,0 +1,268 @@
+
+    auto operation = ArithmeticImmediateInstruction::Operation::Add;
+    switch (raw_parts.funct3) {
+    case 0b000:
+        operation = ArithmeticImmediateInstruction::Operation::Add;
+        break;
+    case 0b010:
+        operation = ArithmeticImmediateInstruction::Operation::SetLessThan;
+        break;
+    case 0b011:
+        operation = ArithmeticImmediateInstruction::Operation::SetLessThanUnsigned;
+        break;
+    case 0b100:
+        operation = ArithmeticImmediateInstruction::Operation::Xor;
+        break;
+    case 0b110:
+        operation = ArithmeticImmediateInstruction::Operation::Or;
+        break;
+    case 0b111:
+        operation = ArithmeticImmediateInstruction::Operation::And;
+        break;
+    case 0b001:
+        operation = ArithmeticImmediateInstruction::Operation::ShiftLeftLogical;
+        break;
+    case 0b101:
+        if ((instruction & (1 << 30)) == 0)
+            operation = ArithmeticImmediateInstruction::Operation::ShiftRightLogical;
+        else
+            operation = ArithmeticImmediateInstruction::Operation::ShiftRightArithmetic;
+        // Delete the possibly set 10th bit that distinguishes SRLI and SRAI.
+        raw_parts.imm &= ~(1 << 10);
+        break;
+    default:
+        VERIFY_NOT_REACHED();
+    }
+    return adopt_own(*new (nothrow) ArithmeticImmediateInstruction(operation, raw_parts.imm, raw_parts.rs1, raw_parts.rd));
+}
+
+NonnullOwnPtr<ArithmeticImmediateInstruction> parse_op_imm_32(u32 instruction)
+{
+    auto raw_parts = RawIType::parse(instruction);
+
+    auto operation = ArithmeticImmediateInstruction::Operation::Add;
+    switch (raw_parts.funct3) {
+    case 0b000:
+        operation = ArithmeticImmediateInstruction::Operation::AddWord;
+        break;
+    case 0b001:
+        operation = ArithmeticImmediateInstruction::Operation::ShiftLeftLogicalWord;
+        break;
+    case 0b101:
+        if ((instruction & (1 << 30)) == 0)
+            operation = ArithmeticImmediateInstruction::Operation::ShiftRightLogicalWord;
+        else
+            operation = ArithmeticImmediateInstruction::Operation::ShiftRightArithmeticWord;
+        // Delete the possibly set 10th bit that distinguishes SRLIW and SRAIW.
+        raw_parts.imm &= ~(1 << 10);
+        break;
+    default:
+        VERIFY_NOT_REACHED();
+    }
+    return adopt_own(*new (nothrow) ArithmeticImmediateInstruction(operation, raw_parts.imm, raw_parts.rs1, raw_parts.rd));
+}
+
+NonnullOwnPtr<ArithmeticInstruction> parse_op(u32 instruction)
+{
+    auto raw_parts = RawRType::parse(instruction);
+    // Distinguishes a few closely related operations, like add/sub.
+    auto mode_switch = (raw_parts.funct7 & 0b0100000) == 0;
+    auto is_m_extension = (raw_parts.funct7 & 1) == 1;
+
+    auto operation = ArithmeticInstruction::Operation::Add;
+    if (!is_m_extension) {
+        switch (raw_parts.funct3) {
+        case 0b000:
+            if (mode_switch)
```

Comment:
nit: I find it a bit confusing that `mode_switch` is true if bit 30 is 0. I would invert this personally.